### PR TITLE
Collect metrics on email replies and alias creation mpp 1279

### PR DESCRIPTION
**Pull request by say-yawn**
# About this PR
To ensure that we do not let our product be used for potential abuse we want to track and limit usage of creating alias and responding.

# Acceptance Criteria
- [ ] Add log with condition if replies/day or alias created/day is greater 100 log user
- added a field `last_account_flagged` with date and time of the last time user went above the limit
- added a log to console to be used by secops
- [ ] Add metric for Abuse Alert on Grafana
- completed by DB query
- [ ] Pause alias generation or replies when abuse is detected
- completed by checking to see when user was `last_account_flagged` and if current is greater than `last_account_flagged` and `PREMIUM_FEATURE_PAUSED_DAYS` blocks alias creation or replies

#
Resolves [MPP-1279](https://mozilla-hub.atlassian.net/browse/MPP-1279)

### CODE REVIEW

File: emails/models.py [hunk 1]
```diff
@@ -253,6 +254,54 @@ def add_subdomain(self, subdomain):
...
+            datetime.now(timezone.utc).date(), datetime.min.time()
+        ).astimezone(timezone.utc)
+        midnight_utc_tomorow = datetime.combine(
+            datetime.now(timezone.utc).date(), datetime.min.time()
+        ).astimezone(timezone.utc) + timedelta(days=1)
+        abuse_metric = self.user.abusemetrics_set.filter(
+            first_recorded__gte=midnight_utc_today, first_recorded__lt=midnight_utc_tomorow
+        ).first()
+        if not abuse_metric:
+            abuse_metric = AbuseMetrics.objects.create(user=self.user)
```
Comment (groovecoder) [inline emails/models.py:270]: While we're here, could we clean up old `AbuseMetrics` records we don't need anymore? E.g.,
```
AbuseMetrics.objects.delete(first_recorded__lt=midnight_utc_today)
```

File: emails/models.py [hunk 2]
```diff
@@ -253,6 +254,54 @@ def add_subdomain(self, subdomain):
...
+                self.last_account_flagged +
+                timedelta(days=settings.PREMIUM_FEATURE_PAUSED_DAYS)
+            )
+            if datetime.now(timezone.utc) > account_premium_feature_resumed:
+                # premium feature has been resumed
+                return False
+            # user was flagged and the premiume feature pause period is not yet over
+            return True
+        # user was never flagged
+        return False
```
Comment (groovecoder) [inline emails/models.py:303]: Bit of refactor to optimize ...

```suggestion
    def is_flagged(self):
        if not self.last_account_flagged:
            return False
        account_premium_feature_resumed = (
            self.last_account_flagged +
            timedelta(days=settings.PREMIUM_FEATURE_PAUSED_DAYS)
        )
        if datetime.now(timezone.utc) > account_premium_feature_resumed:
            # premium feature has been resumed
            return False
        # user was flagged and the premiume feature pause period is not yet over
        return True
```

**Review by groovecoder — COMMENTED** [review]
A few suggestions. Overall looking good.

File: emails/models.py [hunk 3]
```diff
@@ -253,6 +254,54 @@ def add_subdomain(self, subdomain):
...
 
+    def update_abuse_metric(self, address_created=False, replied=False):
+        #  TODO: this should be wrapped in atomic to ensure race conditions are properly handled
+        # look for abuse metrics created on the same UTC date, regardless of time.
+        midnight_utc_today = datetime.combine(
+            datetime.now(timezone.utc).date(), datetime.min.time()
+        ).astimezone(timezone.utc)
+        midnight_utc_tomorow = datetime.combine(
+            datetime.now(timezone.utc).date(), datetime.min.time()
+        ).astimezone(timezone.utc) + timedelta(days=1)
```
[the remark this answers is not in the document]
Comment (say-yawn) [reply emails/models.py:265]: Yup. Laughed at my own dumb butt.

File: emails/views.py [hunk 4]
```diff
@@ -430,13 +430,17 @@ def _reply_allowed(from_address, to_address, reply_record):
...
         (from_address == reply_record_email) or
         (stripped_from_address == stripped_reply_record_address)
     ):
         # This is a Relay user replying to an external sender;
         # verify they are premium
+        [from_local_portion, from_domain_portion] = stripped_from_address.split('@')
+        address = _get_address(
+                stripped_from_address, from_local_portion, from_domain_portion
+            )
+        user_profile = address.user.profile_set.first()
```
Comment (groovecoder) [inline emails/views.py:443]: Why is this here? It's breaking replies.

**Review by groovecoder — CHANGES_REQUESTED** [review]

File: emails/views.py (hunk 4 above)
Comment (groovecoder) [reply emails/views.py:443]: This gets the `is_flagged` value in a way that keeps replies working:
```
diff --git a/emails/models.py b/emails/models.py
index 27bdb5c..7540c72 100644
--- a/emails/models.py
+++ b/emails/models.py
@@ -624,10 +624,18 @@ class Reply(models.Model):
     def address(self):
         return self.relay_address or self.domain_address

+    @property
+    def profile(self):
+        return self.address.user.profile_set.first()
+
     @property
     def owner_has_premium(self):
-        profile = self.address.user.profile_set.first()
-        return profile.has_premium
+        return self.profile.has_premium
+
+    @property
+    def owner_is_flagged(self):
+        return self.profile.is_flagged
+


 class AbuseMetrics(models.Model):
diff --git a/emails/views.py b/emails/views.py
index 9ef3a04..d96ec73 100644
--- a/emails/views.py
+++ b/emails/views.py
@@ -436,12 +436,10 @@ def _reply_allowed(from_address, to_address, reply_record):
     ):
         # This is a Relay user replying to an external sender;
         # verify they are premium
-        [from_local_portion, from_domain_portion] = stripped_from_address.split('@')
-        address = _get_address(
-                stripped_from_address, from_local_portion, from_domain_portion
-            )
-        user_profile = address.user.profile_set.first()
-        if reply_record.owner_has_premium and not user_profile.is_flagged:
+        if (
+            reply_record.owner_has_premium and
+            not reply_record.owner_is_flagged
+        ):
             # TODO: send the user an email
             # that replies are a premium feature
             return True
```

File: emails/views.py [hunk 5]
```diff
@@ -517,23 +525,26 @@ def _get_domain_address(local_portion, domain_portion):
...
-        domain_address = DomainAddress.objects.filter(
-            user=user_profile.user, address=local_portion, domain=domain_numerical
-        ).first()
-        if domain_address is None:
-            # TODO: We may want to consider flows when
-            # a user generating alias on a fly was unable to
-            # receive an email due to the following exceptions
-            domain_address = DomainAddress.make_domain_address(
-                user_profile, local_portion, True
+        with transaction.atomic:
```
Comment (groovecoder) [inline emails/views.py:528]:
```suggestion
        with transaction.atomic():
```

**Review by groovecoder — CHANGES_REQUESTED** [review]

**Review by groovecoder — APPROVED** [review]

**Pull request merged**